### ISSUE TITLE
Compressing non-square images produces incorrect (and unreadable) files

### ISSUE THREAD

**Issue by keastrid** [nom-tam-fits/nom-tam-fits#268]
### Discussed in https://github.com/nom-tam-fits/nom-tam-fits/discussions/263 (just realized this should probably be an issue, as outputing junk images probably isn't expected even if the data/setup is wrong)

<div type='discussions-op-text'>

<sup>Originally posted by **keastrid** March  6, 2022</sup>
Hello, 

It seems this is related to #129, except that HCOMPRESS also seems to have issues.

I'm looking for some help compressing images. When trying to compress a non-square image, a file is produced that cannot be read (see error).

I've played around with it for a while, and I can't figure out what I'm doing wrong. Attached is a test image and the bad compressed version of it.

I am attempting to compress with this code:
```
try (var f = new Fits()) {
	var stack = imp.getStack();
	var ip = stack.getProcessor(slice);
	var type = ImageType.getType(ip);
	var data = type.make2DArray(ip);//todo handle color images
	var hdu = Fits.makeHDU(data);
	var header = hdu.getHeader();

	var compressedHDU = CompressedImageHDU.fromImageHDU((ImageHDU) hdu);//Tried with , -1, -1 as well

	// To have lossless compression, we must handle floating point and integer separately
	if (type.isFloatingPoint()) {
		//
	} else {
		/*compressedHDU.setCompressAlgorithm(Compression.ZCMPTYPE_RICE_1)//RICE also fails
				.setQuantAlgorithm((String) null)
				.getCompressOption(RiceCompressOption.class)
				.setBlockSize(32)
				.setBytePix(2);*/
		compressedHDU.setCompressAlgorithm(Compression.ZCMPTYPE_HCOMPRESS_1)//
				.setQuantAlgorithm((String) null)//
				.getCompressOption(HCompressorOption.class)//
				/**/.setSmooth(true)//
				/**/.setScale(1);
	}

	// Add compressed HDU
	compressedHDU.compress();
	f.addHDU(compressedHDU);

	Files.createDirectories(Path.of(path).getParent());
	if (!Path.of(path).toFile().exists()) Files.createFile(Path.of(path));

	f.write(Path.of(path).toFile());
} catch (IOException | FitsException e) {
	e.printStackTrace();
}
```

Error:
```java.lang.ArrayIndexOutOfBoundsException: Index 4096 out of bounds for length 4096
	at nom.tam.image.compression.tile.TileDecompressorInitialisation.init(TileDecompressorInitialisation.java:72)
	at nom.tam.image.compression.tile.TileDecompressorInitialisation.init(TileDecompressorInitialisation.java:41)
	at nom.tam.image.tile.operation.AbstractTiledImageOperation.createTiles(AbstractTiledImageOperation.java:128)
	at nom.tam.image.compression.tile.TiledImageCompressionOperation.read(TiledImageCompressionOperation.java:195)
	at nom.tam.image.compression.hdu.CompressedImageData.getUncompressedData(CompressedImageData.java:88)
	at nom.tam.image.compression.hdu.CompressedImageHDU.getUncompressedData(CompressedImageHDU.java:171)
	at nom.tam.image.compression.hdu.CompressedImageHDU.asImageHDU(CompressedImageHDU.java:131)
```

[CompressTest.zip](https://github.com/nom-tam-fits/nom-tam-fits/files/8193199/CompressTest.zip)
</div>

**Comment by attipaci** [comment]
This is probably the same bug as #129. Can you see if the two are indeed related.

The compression support to the library was added by Ritchie, who has been largely inactive for a number of years now. I have very little understanding of the work he did or how compression was/is supposed to be working. We clearly have a need for someone to pick up from where Ritchie left off... 

I don't think I will be able to tackle this, at least not in the foreseeable future. I'm sorry it's causing you issues. I hope you find a way around it, or perhaps a fix for the underlying problem itself?

I wish I had a better answer. I'm afraid I really don't.

**Comment by attipaci** [comment]
PS. My poor man's approach for compression is the old fashioned way of just GZIP-ing the whole thing. E.g. as:

```java
  String fileName = "some.fits.gz";   // The file you want to write compressed FITS to...
  Fits fits;                          // The Fits data object

  // Create a GZIP-ed FITS output stream...
  FitsOutputStream out = new FitsOutputStream(new GZIPOutputStream(new FileOutputStream(new File(fileName))));

  // Write the FITS gzipped.
  fits.write(out);
  out.close();
``` 

This works also if you add HDUs incrementally, and it is bullet proof. I generally don't see much need for per-HDU compression support...

**Comment by keastrid** [comment]
Yes, I do think #129 is related.

I've tried looking into it myself for AstroImageJ*, but I also can't really make sense of how compression works here. I do agree per-HDU compression is silly, but the ratios are nice in comparison to gzip. If I get some time I may try figuring it out, its really the type of API that should have a a "data in -> lossless out" default path as well without the nitty-gritty compression options.

** We've since started to use our own fork, to add support for our nonstandard keyword usage and to fix issues since development seemed to stop here. I *am* recreating these issues with "vanilla" nom.tam (and so far the only change has been keyword support). We are GPL, but I have no problem relicensing the nom.tam changes if you're interested in any present or future changes we make (since this is active again I'll probably try and upstream most changes that aren't specific to us anyways). Just didn't want this to come up as a surprise later on :)

**Comment by attipaci** [comment]
Yes please. It would be great to reintegrate your improvements to mainline. I'd much prefer if we had one good Java FITS library than having it be fragmented into a number of forks, each with different fixes and features -- which is what has been happening when nom-tam-fits went unmaintained for a number of years. If you can find and fix the non-square image compression bug that would be stellar.

Feel free to create pull requests. I'll promise to look at them without too much delay, give feedback, and/or merge when ready. Apart from not breaking existing test cases (if possible), I will just ask you to add relevant test cases to all new lines of code. (Sometimes breaking tests is also OK, because some of the existing tests were not well set up in the first case.)

This library is 100% in community hands. I'm just a contributor too. I stepped up to overseeing the lib only because I got tired of it going unmaintained for so long. Thanks in advance for chipping in. :-)

**Comment by rmathar** [comment]
The key advantage of per-HDU compression is that one can read all HDU keywords (with all the advantage of random-file access) without touching or uncompressing the data. In the file-compressed variant there is no need to uncompress huge amounts of data just because they are a single serialized compressed stream. This is in particular useful if one generates databases of header-keyword values of many FITS files with many HDU's.

**Comment by attipaci** [comment]
@keastrid, I think I got this issue fixed in the above pull request (#318). I used the test image you provided to verify that the image can be compressed -> writtern -> read back -> uncompressed, and result in the the same image as before. This is true both with the default tiling, and with various different tile sizes -- which need not be divisors of the original image size. I will merge this fix within days so it can make it into the upcoming 1.17.0 release (around Sep 15). But let me know if there is more to be desired on this issue...
